relay-cogs/src/cogs.rs

```diff
@@ -0,0 +1,336 @@
+use core::fmt;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+use std::time::Instant;
+use std::usize;
+
+use crate::{utils::EnumMap, AppFeature, ResourceId};
+use crate::{CogsMeasurement, CogsRecorder, Value};
+
+/// COGS measurements collector.
+///
+/// The collector is cheap to clone.
+#[derive(Clone)]
+pub struct Cogs {
+    recorder: Arc<dyn CogsRecorder>,
+}
+
+impl Cogs {
+    /// Creates a new [`Cogs`] from a [`recorder`](CogsRecorder).
+    pub fn new<T>(recorder: T) -> Self
+    where
+        T: CogsRecorder + 'static,
+    {
+        Self {
+            recorder: Arc::new(recorder),
+        }
+    }
+
+    /// Shortcut for creating a [`Cogs`] from a [`crate::NoopRecorder`].
+    ///
+    /// All collected measurements will be dropped.
+    pub fn noop() -> Self {
+        Self::new(crate::NoopRecorder)
+    }
+}
+
+impl Cogs {
+    /// Starts a recording for a COGS measurement.
+    ///
+    /// When the returned token is dropped the measurement will be recorded
+    /// with the configured [recorder](CogsRecorder).
+    ///
+    /// The recorded measurement can be attributed to multiple features by supplying a
+    /// weighted [`AppFeatures`]. A single [`AppFeature`] attributes the entire measurement
+    /// to the feature.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// # use relay_cogs::{AppFeature, Cogs, ResourceId};
+    /// # struct Span;
+    /// # fn scrub_sql(_: &mut Span) {}
+    /// # fn extract_tags(_: &mut Span) {};
+    ///
+    /// fn process_span(cogs: &Cogs, span: &mut Span) {
+    ///     let _token = cogs.timed(ResourceId::Relay, AppFeature::Spans);
+    ///
+    ///     scrub_sql(span);
+    ///     extract_tags(span);
+    /// }
+    ///
+    /// ```
+    pub fn timed<F: Into<AppFeatures>>(&self, resource: ResourceId, features: F) -> CogsToken {
```

Comment:
Any reason to go with the token approach over a closure? I find the increased readability worth the decreased flexibility. Something like the following:
```rs
Cogs::timed(some_features, || {
  // ...
});
```

Reply:
An additional indentation level hurts readability, your example is actually also missing a `move` and then internally the closure still needs to use a token to correctly handle all cases including panics.